[CI] BlockHashRandomizedTests test {forcePackedHash=false groups=1 maxValuesPerPosition=1 dups=0 allowedTypes=[Ordinals[dictionarySize=10]]} failing

**Build Scans:**
- [elasticsearch-periodic-platform-support #4486 / amazonlinux-2023_platform-support-aws](https://gradle-enterprise.elastic.co/s/xltobyk5k2dfe)
- [elasticsearch-periodic-platform-support #4475 / amazonlinux-2023_platform-support-aws](https://gradle-enterprise.elastic.co/s/ktui43jkjiw56)
- [elasticsearch-periodic #4434 / graalvm-ce17_checkpart3_java-matrix](https://gradle-enterprise.elastic.co/s/nwdaumraczhvu)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:esql:compute:test" --tests "org.elasticsearch.compute.aggregation.blockhash.BlockHashRandomizedTests.test {forcePackedHash=false groups=1 maxValuesPerPosition=1 dups=0 allowedTypes=[Ordinals[dictionarySize=10]]}" -Dtests.seed=8C5111DC7877EEB3 -Dtests.locale=en-MO -Dtests.timezone=Africa/Ceuta -Druntime.java=22
```

**Applicable branches:**
8.x

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('test%20%7BforcePackedHash%3Dfalse%20groups%3D1%20maxValuesPerPosition%3D1%20dups%3D0%20allowedTypes%3D%5BOrdinals%5BdictionarySize%3D10%5D%5D%7D'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.compute.aggregation.blockhash.BlockHashRandomizedTests'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure Message:**
```
java.lang.AssertionError: Expected a list containing
 0: a list containing
   0: <[41 75 51 77 78]>
 1: a list containing
   0: <[43 4f 50 76 4f]>
 2: a list containing
   0: <[43 57 69 55 73]>
 3: a list containing
   0: <[43 62 77 4d 77]>
 4: a list containing
   0: <[43 7a 47 74 6d]>
 5: a list containing
   0: <[44 43 44 73 49]>
 6: a list containing
   0: <[44 55 48 64 6c]>
 7: a list containing
   0: <[44 63 46 44 58]>
 8: a list containing
   0: <[46 4d 48 59 79]>
 9: a list containing
   0: <[46 51 77 47 53]>
10: a list containing
   0: <[47 69 72 56 6a]>
11: a list containing
   0: <[48 50 53 45 56]>
12: a list containing
   0: <[48 74 75 4e 73]>
13: a list containing
   0: <[49 62 52 7a 62]>
14: a list containing
   0: <[4a 54 66 4c 78]>
15: a list containing
   0: <[4a 70 49 78 69]>
16: a list containing
   0: <[4b 47 51 41 52]>
17: a list containing
   0: <[4c 41 43 50 55]>
18: a list containing
   0: <[4c 46 70 6d 73]>
19: a list containing
   0: <[4f 48 66 51 4d]>
20: a list containing
   0: <[50 45 4f 51 61]>
21: a list containing
   0: <[50 4e 50 6a 57]>
22: a list containing
   0: <[50 4f 78 71 54]>
23: a list containing
   0: <[52 49 48 41 79]>
24: a list containing
   0: <[52 4b 69 7a 76]>
25: a list containing
   0: <[52 50 65 49 65]>
26: a list containing
   0: <[52 57 47 77 63]>
27: a list containing
   0: expected <[52 74 62 69 45]> but was <[52 6d 6f 58 55]>
28: a list containing
   0: expected <[53 57 48 57 49]> but was <[52 74 62 69 45]>
29: a list containing
   0: expected <[54 4c 55 51 63]> but was <[53 57 48 57 49]>
30: a list containing
   0: expected <[55 4d 79 50 54]> but was <[54 4c 55 51 63]>
31: a list containing
   0: expected <[55 55 67 56 57]> but was <[55 4d 79 50 54]>
32: a list containing
   0: expected <[56 45 43 53 6a]> but was <[55 55 67 56 57]>
33: a list containing
   0: expected <[56 6a 7a 74 45]> but was <[56 45 43 53 6a]>
34: a list containing
   0: expected <[59 6b 4f 4e 69]> but was <[56 6a 7a 74 45]>
35: a list containing
   0: expected <[59 6e 64 42 72]> but was <[59 6b 4f 4e 69]>
36: a list containing
   0: expected <[5a 4b 6d 58 4c]> but was <[59 6e 64 42 72]>
37: a list containing
   0: expected <[5a 7a 55 57 59]> but was <[5a 4b 6d 58 4c]>
38: a list containing
   0: expected <[61 47 41 67 69]> but was <[5a 7a 55 57 59]>
39: a list containing
   0: expected <[61 59 73 59 44]> but was <[61 47 41 67 69]>
40: a list containing
   0: expected <[61 68 6e 4d 51]> but was <[61 59 73 59 44]>
41: a list containing
   0: expected <[61 6f 43 48 46]> but was <[61 68 6e 4d 51]>
42: a list containing
   0: expected <[62 7a 6d 67 59]> but was <[61 6f 43 48 46]>
43: a list containing
   0: expected <[63 48 56 61 46]> but was <[62 7a 6d 67 59]>
44: a list containing
   0: expected <[63 5a 73 77 4e]> but was <[63 48 56 61 46]>
45: a list containing
   0: expected <[63 72 73 72 73]> but was <[63 5a 73 77 4e]>
46: a list containing
   0: expected <[65 62 6d 62 58]> but was <[63 72 73 72 73]>
47: a list containing
   0: expected <[65 6a 71 70 6e]> but was <[65 62 6d 62 58]>
48: a list containing
   0: expected <[66 47 69 74 44]> but was <[65 6a 71 70 6e]>
49: a list containing
   0: expected <[66 4c 49 69 67]> but was <[66 47 69 74 44]>
50: a list containing
   0: expected <[67 58 49 57 45]> but was <[66 4c 49 69 67]>
51: a list containing
   0: expected <[67 77 50 46 6f]> but was <[67 58 49 57 45]>
52: a list containing
   0: expected <[68 56 43 66 50]> but was <[67 77 50 46 6f]>
53: a list containing
   0: expected <[68 64 52 75 6d]> but was <[68 56 43 66 50]>
54: a list containing
   0: expected <[68 6b 46 5a 6a]> but was <[68 64 52 75 6d]>
55: a list containing
   0: expected <[69 4a 44 4e 57]> but was <[68 6b 46 5a 6a]>
56: a list containing
   0: expected <[69 66 61 45 6d]> but was <[69 4a 44 4e 57]>
57: a list containing
   0: expected <[69 67 7a 4f 6e]> but was <[69 66 61 45 6d]>
58: a list containing
   0: expected <[69 68 67 4d 46]> but was <[69 67 7a 4f 6e]>
59: a list containing
   0: expected <[6a 65 4e 46 74]> but was <[69 68 67 4d 46]>
60: a list containing
   0: expected <[6a 74 75 68 47]> but was <[6a 65 4e 46 74]>
61: a list containing
   0: expected <[6b 55 6c 6e 6f]> but was <[6a 74 75 68 47]>
62: a list containing
   0: expected <[6d 68 45 6a 73]> but was <[6b 55 6c 6e 6f]>
63: a list containing
   0: expected <[6e 63 43 67 68]> but was <[6d 68 45 6a 73]>
64: a list containing
   0: expected <[6f 51 59 46 4d]> but was <[6e 63 43 67 68]>
65: a list containing
   0: expected <[70 4f 69 74 45]> but was <[6f 51 59 46 4d]>
66: a list containing
   0: expected <[70 66 55 76 79]> but was <[70 4f 69 74 45]>
67: a list containing
   0: expected <[70 76 55 54 46]> but was <[70 66 55 76 79]>
68: a list containing
   0: expected <[70 79 62 43 78]> but was <[70 76 55 54 46]>
69: a list containing
   0: expected <[71 74 62 6b 41]> but was <[70 79 62 43 78]>
70: a list containing
   0: expected <[72 43 6c 6d 65]> but was <[71 74 62 6b 41]>
71: a list containing
   0: expected <[74 4d 64 73 73]> but was <[72 43 6c 6d 65]>
72: a list containing
   0: expected <[74 63 51 57 75]> but was <[74 4d 64 73 73]>
73: a list containing
   0: expected <[74 6e 4c 65 6f]> but was <[74 63 51 57 75]>
74: a list containing
   0: expected <[76 6f 65 4b 4b]> but was <[74 6e 4c 65 6f]>
75: a list containing
   0: expected <[77 68 69 66 6c]> but was <[76 6f 65 4b 4b]>
76: a list containing
   0: expected <[79 59 4d 6e 67]> but was <[77 68 69 66 6c]>
77: a list containing
   0: expected <[7a 58 58 52 42]> but was <[79 59 4d 6e 67]>
78: a list containing
   0: expected <[7a 7a 62 5a 49]> but was <[7a 58 58 52 42]>
79: <unexpected> but was <[[7a 7a 62 5a 49]]>
```

**Issue Reasons:**
- [8.x] 3 failures in test test {forcePackedHash=false groups=1 maxValuesPerPosition=1 dups=0 allowedTypes=[Ordinals[dictionarySize=10]]} (0.5% fail rate in 657 executions)
- [8.x] 2 failures in step amazonlinux-2023_platform-support-aws (12.5% fail rate in 16 executions)
- [8.x] 2 failures in pipeline elasticsearch-periodic-platform-support (11.8% fail rate in 17 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/blockhash/BlockHashRandomizedTests.java
+++ b/x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/blockhash/BlockHashRandomizedTests.java
@@ -475,17 +475,14 @@ public class BlockHashRandomizedTests extends ESTestCase {
 
         @Override
         public BasicBlockTests.RandomBlock randomBlock(int positionCount, int maxValuesPerPosition, int dups) {
-            List<Map.Entry<String, Integer>> dictionary = new ArrayList<>();
+            Map<String, Integer> dictionary = new HashMap<>();
+            Set<String> keys = dictionary(maxValuesPerPosition);
             List<List<Object>> values = new ArrayList<>(positionCount);
             try (
                 IntBlock.Builder ordinals = TestBlockFactory.getNonBreakingInstance()
                     .newIntBlockBuilder(positionCount * maxValuesPerPosition);
                 BytesRefVector.Builder bytes = TestBlockFactory.getNonBreakingInstance().newBytesRefVectorBuilder(maxValuesPerPosition);
             ) {
-                for (String value : dictionary(maxValuesPerPosition)) {
-                    bytes.appendBytesRef(new BytesRef(value));
-                    dictionary.add(Map.entry(value, dictionary.size()));
-                }
                 for (int p = 0; p < positionCount; p++) {
                     int valueCount = between(1, maxValuesPerPosition);
                     int dupCount = between(0, dups);
@@ -497,10 +494,14 @@ public class BlockHashRandomizedTests extends ESTestCase {
                         ordinals.beginPositionEntry();
                     }
                     for (int v = 0; v < valueCount; v++) {
-                        Map.Entry<String, Integer> value = randomFrom(dictionary);
-                        valuesAtPosition.add(new BytesRef(value.getKey()));
-                        ordinals.appendInt(value.getValue());
-                        ordsAtPosition.add(value.getValue());
+                        String key = randomFrom(keys);
+                        int ordinal = dictionary.computeIfAbsent(key, k -> {
+                            bytes.appendBytesRef(new BytesRef(k));
+                            return dictionary.size();
+                        });
+                        valuesAtPosition.add(new BytesRef(key));
+                        ordinals.appendInt(ordinal);
+                        ordsAtPosition.add(ordinal);
                     }
                     for (int v = 0; v < dupCount; v++) {
                         ordinals.appendInt(randomFrom(ordsAtPosition));

